src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -168,29 +196,88 @@ struct TensorContract
 
 /*!
  * \ingroup TensorExpressionsGroup
+ * \brief Returns the values of the first TensorIndexs to contract in an
+ * expression
+ *
+ * \details Given a list of values that represent an expression's generic index
+ * encodings, this function looks to see if it can find a pair of values that
+ * encode one generic index and the generic index with opposite valence, such as
+ * `ti_A_t` and `ti_a_t`. This denotes a pair of indices that will need to be
+ * contracted. If there exists more than one such pair of indices in the
+ * expression, the first pair of values found will be returned.
+ *
+ * For example, if we have tensor \f${R^{ab}}_{ab}\f$ represented by the tensor
+ * expression, `R(ti_A, ti_B, ti_a, ti_b)`, then this will return the pair of
+ * values encoding `ti_A_t` and `ti_a_t`.
+ *
+ * @param tensorindex_values the TensorIndex values of a tensor expression
+ * @return the first pair of TensorIndex values to contract
  */
-template <size_t ReplacedArg1, size_t ReplacedArg2, typename T, typename X,
-          typename Symm, typename IndexList, typename Args>
-SPECTRE_ALWAYS_INLINE auto contract(
-    const TensorExpression<T, X, Symm, IndexList, Args>& t) {
-  return TensorContract<tmpl::size_t<ReplacedArg1>,
-                        tmpl::size_t<ReplacedArg2>, T, X, Symm, IndexList,
-                        Args>(~t);
+template <size_t NumIndices>
+SPECTRE_ALWAYS_INLINE static constexpr std::pair<size_t, size_t>
+get_first_tensorindex_values_to_contract(
+    const std::array<size_t, NumIndices>& tensorindex_values) noexcept {
+  for (size_t i = 0; i < tensorindex_values.size(); ++i) {
+    const size_t current_value = gsl::at(tensorindex_values, i);
+    const size_t opposite_value_to_find =
+        get_tensorindex_value_with_opposite_valence(current_value);
+    for (size_t j = i + 1; j < tensorindex_values.size(); ++j) {
+      if (opposite_value_to_find == gsl::at(tensorindex_values, j)) {
+        // We found both the lower and upper version of a generic index in the
+        // list of generic indices, so we return this pair
+        return std::pair{current_value, opposite_value_to_find};
```

Comment:
[optional] what if you did the check here to return `upper, lower`? Then you don't need the check in the Contract struct.

Reply:
Done! (Except, instead of the upper valence index's position and lower valence index's position, I return the smaller position and lower position, regardless of valence)